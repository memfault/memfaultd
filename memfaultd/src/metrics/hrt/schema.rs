//
// Copyright (c) Memfault, Inc.
// See License.txt for details
//! Contains the on disk representation of an HRT report.
//!
//! This module provides the schema for the on disk format of the high resolution telemetry data.
//! It also provides implementations to convert the in memory representation of the data to the on
//! disk format.

use std::io::Write;
use std::{fs::File, io::BufWriter, path::Path, time::Duration};

use chrono::Utc;
use eyre::{Error, Result};
use flate2::write::GzEncoder;
use serde::{Deserialize, Serialize};
use serde_json::to_writer;

use super::report::HrtReport;
use crate::{
    build_info::VERSION,
    mar::{CompressionAlgorithm, MarEntry, MarEntryBuilder, Metadata},
    metrics::{KeyedMetricReading, MetricReading},
    network::NetworkConfig,
    util::{fs::DEFAULT_GZIP_COMPRESSION_LEVEL, serialization::milliseconds_to_duration},
};

const SCHEMA_VERSION: u32 = 1;
const PRODUCER_ID: &str = "memfaultd";
const MIME_TYPE: &str = "application/vnd.memfault.hrt.v1";
const FILE_NAME: &str = "hrt.json.gz";

#[derive(Serialize, Deserialize, Debug)]
/// Represents the on disk format of the high resolution telemetry data.
pub struct HighResTelemetryV1 {
    schema_version: u32,
    producer: Producer,
    start_time: i64,
    #[serde(rename = "duration_ms", with = "milliseconds_to_duration")]
    duration: Duration,
    rollups: Vec<Rollup>,
}

impl TryFrom<HrtReport> for HighResTelemetryV1 {
    type Error = Error;

    fn try_from(report: HrtReport) -> Result<Self, Self::Error> {
        let now = Utc::now();

        let rollups = report
            .readings
            .into_iter()
            .map(|(key, data)| Rollup {
                metadata: RollupMetadata {
                    string_key: key.to_string(),
                    metric_type: data.metadata.metric_type,
                    data_type: data.metadata.data_type,
                    internal: data.metadata.internal,
                },
                data: data.readings,
            })
            .collect();

        let duration = now - report.start_time;

        Ok(Self {
            schema_version: SCHEMA_VERSION,
            producer: Producer::new(),
            start_time: report.start_time.timestamp_millis(),
            duration: duration.to_std()?,
            rollups,
        })
    }
}

#[cfg(test)]
impl HighResTelemetryV1 {
    /// Helper function to easily sort rollups vector in snapshot tests
    pub fn sort_rollups(&mut self) {
        self.rollups
            .sort_by(|a, b| a.metadata.string_key.cmp(&b.metadata.string_key));
    }
}

#[derive(Serialize, Deserialize, Debug)]
/// A single HRT data point.
pub struct Datum {
    /// Timestamp in milliseconds since the epoch.
    t: i64,
    /// Value of the data point.
    value: f64,
}

impl From<&KeyedMetricReading> for Datum {
    fn from(reading: &KeyedMetricReading) -> Self {
        let (t, value) = match reading.value {
            MetricReading::TimeWeightedAverage {
                timestamp, value, ..
            } => (timestamp.timestamp_millis(), value),
            MetricReading::Histogram {
                timestamp, value, ..
            } => (timestamp.timestamp_millis(), value),
            MetricReading::Counter {
                timestamp, value, ..
            } => (timestamp.timestamp_millis(), value),
            MetricReading::Gauge {
                timestamp, value, ..
            } => (timestamp.timestamp_millis(), value),
            MetricReading::Rssi {
                timestamp, value, ..
            } => (timestamp.timestamp_millis(), value),
            // TODO: Add support for Strings in HRT
            MetricReading::ReportTag { timestamp, .. } => (timestamp.timestamp_millis(), 0.0),
            // TODO: Add support for Booleans in HRT
            MetricReading::Bool { timestamp, .. } => (timestamp.timestamp_millis(), 0.0),
        };
        Datum { t, value }
    }
}

#[derive(Serialize, Deserialize, Debug)]
/// Metadata for a single metric key.
pub struct RollupMetadata {
    string_key: String,
    metric_type: HrtMetricType,
    data_type: DataType,
    internal: bool,
}

#[derive(Serialize, Deserialize, Debug)]
/// Represents all data for a single metric.
pub struct Rollup {
    metadata: RollupMetadata,
    data: Vec<Datum>,
}

#[derive(Serialize, Deserialize, Debug)]
/// Tells us who produced the data.
pub struct Producer {
    id: String,
    version: String,
}

impl Producer {
    fn new() -> Self {
        Self {
            id: PRODUCER_ID.to_string(),
            version: VERSION.to_string(),
        }
    }
}

#[derive(Serialize, Deserialize, Debug, Clone)]
#[serde(rename_all = "lowercase")]
// TODO: Remove once used
#[allow(dead_code)]
/// Metric types as defined by the HRT schema.
pub enum HrtMetricType {
    Counter,
    Gauge,
    Property,
    Event,
}

#[derive(Serialize, Deserialize, Debug, Clone)]
#[serde(rename_all = "lowercase")]
// TODO: Remove once used
#[allow(dead_code)]
/// Data types as defined by the HRT schema.
pub enum DataType {
    Double,
    String,
    Boolean,
}

/// Writes an HRT report to disk.
///
/// This function serializes the given HRT report to JSON and saves it in a CDR MAR entry.
pub fn write_report_to_disk(
    report: HrtReport,
    mar_path: &Path,
    network_config: &NetworkConfig,
) -> Result<MarEntry> {
    let start_time = report.start_time;
    let hrt = HighResTelemetryV1::try_from(report)?;

    let mar_builder =
        MarEntryBuilder::new(mar_path)?.set_metadata(Metadata::new_custom_data_recording(
            Some(start_time),
            hrt.duration,
            vec![MIME_TYPE.to_string()],
            "hrt".to_string(),
            FILE_NAME.to_string(),
            Some(CompressionAlgorithm::Gzip),
        ));

    let hrt_path = mar_builder.make_attachment_path_in_entry_dir(FILE_NAME);
    let hrt_file = File::create(hrt_path)?;
    let mut gz_encoder = BufWriter::new(GzEncoder::new(hrt_file, DEFAULT_GZIP_COMPRESSION_LEVEL));
    to_writer(&mut gz_encoder, &hrt)?;
    gz_encoder.flush()?;

    mar_builder.save(network_config)
}

#[cfg(test)]
mod test {
    use std::{fs::File, io::BufReader, num::NonZeroU32};

    use super::*;

    use crate::{
        mar::manifest,
        metrics::{
            hrt::{
                report::{HrtReadingData, HrtReadingMetadata},
                HrtReport,
            },
            MetricStringKey,
        },
        network::NetworkConfig,
        util::rate_limiter::RateLimiter,
    };

    use chrono::{TimeZone, Utc};
    use flate2::bufread::GzDecoder;
    use insta::assert_json_snapshot;
    use rstest::rstest;
    use serde_json::json;

    #[test]
    fn test_serialization() {
        let mut report = build_report();

        report.readings.insert(
            "test".into(),
            HrtReadingData {
                readings: vec![Datum {
                    t: 1010,
                    value: 220.0,
                }],
                metadata: HrtReadingMetadata {
                    metric_type: HrtMetricType::Counter,
                    data_type: DataType::Double,
                    internal: false,
                },
            },
        );

        let hrt = HighResTelemetryV1::try_from(report).unwrap();
        assert_json_snapshot!(json!(hrt), {".duration_ms" => 0, ".producer.version" => "1.2.3"});
    }

    #[test]
    fn test_cdr_write() {
        let mut report = build_report();

        report.readings.insert(
            "test".into(),
            HrtReadingData {
                readings: vec![
                    Datum {
                        t: 1234,
                        value: 567.0,
                    },
                    Datum {
                        t: 5678,
                        value: 123.0,
                    },
                ],
                metadata: HrtReadingMetadata {
                    metric_type: HrtMetricType::Counter,
                    data_type: DataType::Double,
                    internal: false,
                },
            },
        );

        let tmp_dir = tempfile::tempdir().unwrap();
        let mar_path = tmp_dir.path();

        let entry = write_report_to_disk(report, mar_path, &NetworkConfig::test_fixture()).unwrap();

        let hrt_filename = match entry.manifest.metadata {
            Metadata::CustomDataRecording {
                recording_file_name,
                ..
            } => recording_file_name,
            _ => panic!("Unexpected metadata type"),
        };

        let hrt_file_path = entry.path.join(hrt_filename);
        let manifest_path = entry.path.join("manifest.json");
        assert!(hrt_file_path.exists());
        assert!(manifest_path.exists());

        let hrt_file = File::open(hrt_file_path).unwrap();
        let gz_decoder = GzDecoder::new(BufReader::new(hrt_file));
        let hrt: HighResTelemetryV1 = serde_json::from_reader(gz_decoder).unwrap();
        assert_json_snapshot!(json!(hrt), {".duration_ms" => 0, ".producer.version" => "1.2.3"});

        let manifest_file = File::open(manifest_path).unwrap();
        let manifest: manifest::Manifest = serde_json::from_reader(manifest_file).unwrap();
        assert_json_snapshot!(manifest.metadata, {".metadata.duration_ms" => 0});
    }

    #[rstest]
    #[case("counter", MetricReading::Counter { value: 123.0, timestamp: Utc.timestamp_millis_opt(1234).unwrap() })]
    #[case("gauge", MetricReading::Gauge { value: 123.0, timestamp: Utc.timestamp_millis_opt(1234).unwrap() })]
    #[case("histogram", MetricReading::Histogram { value: 123.0, timestamp: Utc.timestamp_millis_opt(1234).unwrap() })]
    fn test_datum_conversion(#[case] case: &str, #[case] reading: MetricReading) {
        let metric = KeyedMetricReading {
            name: MetricStringKey::from("test"),
            value: reading,
        };
        let datum = Datum::from(&metric);

        assert_json_snapshot!(case, json!(datum));
    }

    fn build_report() -> HrtReport {
        let start_time = Utc.with_ymd_and_hms(1991, 3, 25, 0, 0, 0).unwrap();
        HrtReport {
            start_time,
            readings: Default::default(),
            rate_limiter: RateLimiter::new(
                NonZeroU32::new(7500).expect("Zero value passed to non-zero constructor"),
            ),
        }
    }
}
