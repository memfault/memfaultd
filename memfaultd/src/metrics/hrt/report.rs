//
// Copyright (c) Memfault, Inc.
// See License.txt for details
//! Contains the in memory representation of an HRT report.
//!
//! The HRT report represents a collection of readings over a given time period. This report
//! differs from the standard metric report in that it keeps all readings over time and does
//! not represent an aggregate.

use std::{
    collections::{hash_map::Entry, HashMap},
    num::NonZeroU32,
};

use chrono::{DateTime, Utc};
use log::{debug, warn};

use super::schema::{DataType, Datum, HrtMetricType};
use crate::{
    metrics::{KeyedMetricReading, MetricReading, MetricStringKey, MetricTimestamp},
    util::rate_limiter::RateLimiter,
};

pub const HRT_DEFAULT_MAX_SAMPLES_PER_MIN: u32 = 750;

/// Represents the in memory format of high resolution telemetry data.
///
/// This is a helper struct that allows for easier access to the data in the report. It
/// differs from the on disk format in that it optimizes for ease of access over a more
/// compact representation.
pub struct HrtReport {
    pub start_time: MetricTimestamp,
    pub readings: HashMap<MetricStringKey, HrtReadingData>,
    pub rate_limiter: RateLimiter<DateTime<Utc>>,
}

impl HrtReport {
    pub fn new(max_readings_per_min: NonZeroU32) -> Self {
        let start_time = Utc::now();

        // Cap the configured rate limit at
        // 2x the default
        let rate_limit_upper_bound = NonZeroU32::new(HRT_DEFAULT_MAX_SAMPLES_PER_MIN * 2)
            .expect("Default HRT limit should be nonzero");
        let rate_limit = if max_readings_per_min > rate_limit_upper_bound {
            debug!(
                "Configured HRT rate limit is {} - capping at {} samples per minute",
                max_readings_per_min, rate_limit_upper_bound
            );
            rate_limit_upper_bound
        } else {
            max_readings_per_min
        };

        Self {
            start_time,
            readings: HashMap::new(),
            rate_limiter: RateLimiter::new(rate_limit),
        }
    }

    pub fn add_metric(&mut self, reading: &KeyedMetricReading) {
        if let Err(e) =
            self.rate_limiter
                .run_within_limits(reading.value.timestamp(), |rate_limited_calls| {
                    if let Some(limited) = rate_limited_calls {
                        debug!("{} HRT readings were rate limited.", limited.count);
                    };
                    match self.readings.entry(reading.name.clone()) {
                        Entry::Occupied(mut o) => {
                            let hrt_data = o.get_mut();
                            hrt_data.push_reading(reading);
                            Ok(())
                        }
                        Entry::Vacant(v) => {
                            v.insert(HrtReadingData::from(reading));
                            Ok(())
                        }
                    }
                })
        {
            warn!("Failed to add HRT metric reading: {}", e);
        }
    }
}

impl Default for HrtReport {
    fn default() -> Self {
        let start_time = Utc::now();

        Self {
            start_time,
            readings: HashMap::new(),
            rate_limiter: RateLimiter::new(
                // Default
                NonZeroU32::new(HRT_DEFAULT_MAX_SAMPLES_PER_MIN)
                    .expect("Zero value passed to non-zero constructor"),
            ),
        }
    }
}

#[derive(Debug)]
/// Represents both reading data and metadata for a single metric key.
pub struct HrtReadingData {
    pub readings: Vec<Datum>,
    pub metadata: HrtReadingMetadata,
}

impl From<&KeyedMetricReading> for HrtReadingData {
    fn from(reading: &KeyedMetricReading) -> Self {
        Self {
            readings: vec![Datum::from(reading)],
            metadata: HrtReadingMetadata::from(reading),
        }
    }
}

impl HrtReadingData {
    fn push_reading(&mut self, reading: &KeyedMetricReading) {
        self.readings.push(Datum::from(reading))
    }
}

#[derive(Debug)]
/// Metadata for a single metric key.
pub struct HrtReadingMetadata {
    pub metric_type: HrtMetricType,
    pub data_type: DataType,
    pub internal: bool,
}

impl From<&KeyedMetricReading> for HrtReadingMetadata {
    fn from(reading: &KeyedMetricReading) -> Self {
        let (metric_type, data_type) = match reading.value {
            MetricReading::TimeWeightedAverage { .. } => (HrtMetricType::Gauge, DataType::Double),
            MetricReading::Histogram { .. } => (HrtMetricType::Gauge, DataType::Double),
            MetricReading::Counter { .. } => (HrtMetricType::Counter, DataType::Double),
            MetricReading::Gauge { .. } => (HrtMetricType::Gauge, DataType::Double),
            MetricReading::Rssi { .. } => (HrtMetricType::Gauge, DataType::Double),
            MetricReading::ReportTag { .. } => (HrtMetricType::Property, DataType::String),
            MetricReading::Bool { .. } => (HrtMetricType::Property, DataType::Boolean),
        };

        Self {
            metric_type,
            data_type,
            internal: false,
        }
    }
}

#[cfg(test)]
mod tests {
    use crate::metrics::hrt::schema::HighResTelemetryV1;

    use super::*;
    use insta::{assert_json_snapshot, with_settings};
    use rstest::rstest;

    #[rstest]
    fn test_build_hrt_report() {
        let mut hrt_report = HrtReport::new(NonZeroU32::new(2000).unwrap());

        for i in 0..100 {
            hrt_report.add_metric(&KeyedMetricReading::new_counter(
                MetricStringKey::from("test_counter"),
                i as f64,
            ))
        }

        for i in 1000..1100 {
            hrt_report.add_metric(&KeyedMetricReading::new_counter(
                MetricStringKey::from("test_counter_2"),
                i as f64,
            ))
        }

        for i in 500..1100 {
            hrt_report.add_metric(&KeyedMetricReading::new_histogram(
                MetricStringKey::from("test_histo"),
                i as f64,
            ))
        }

        for i in 200..330 {
            hrt_report.add_metric(&KeyedMetricReading::new_gauge(
                MetricStringKey::from("test_gauge"),
                i as f64,
            ))
        }

        let mut hrt_report_serialized = HighResTelemetryV1::try_from(hrt_report).unwrap();
        hrt_report_serialized.sort_rollups();
        with_settings!({sort_maps => true}, {
            assert_json_snapshot!(hrt_report_serialized,
                                  {".producer.version" => "[version]", ".start_time" => "[start_time]", ".rollups[].data[].t" => "[timestamp]", ".duration_ms" => "[duration]"});
        });
    }

    #[rstest]
    fn test_hrt_report_rate_limiting() {
        let mut hrt_report = HrtReport::default();

        for i in 0..1000 {
            hrt_report.add_metric(&KeyedMetricReading::new_counter(
                MetricStringKey::from("test_counter"),
                i as f64,
            ))
        }

        assert_eq!(
            hrt_report
                .readings
                .get(&MetricStringKey::from("test_counter"))
                .unwrap()
                .readings
                .len(),
            750
        );
    }
}
