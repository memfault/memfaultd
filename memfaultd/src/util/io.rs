//
// Copyright (c) Memfault, Inc.
// See License.txt for details
use std::io::{
    self, copy, sink, BufReader, Chain, Cursor, ErrorKind, IoSlice, Read, Seek, SeekFrom, Write,
};

/// A trait for getting the length of a stream.
/// Note std::io::Seek also has a stream_len() method, but that method is fallible.
pub trait StreamLen {
    /// Gets the length of the stream in bytes.
    fn stream_len(&self) -> u64;
}

impl<R: StreamLen> StreamLen for BufReader<R> {
    fn stream_len(&self) -> u64 {
        self.get_ref().stream_len()
    }
}

impl<A: AsRef<[u8]>, B: StreamLen> StreamLen for Chain<Cursor<A>, B> {
    fn stream_len(&self) -> u64 {
        let (a, b) = self.get_ref();

        a.get_ref().as_ref().len() as u64 + b.stream_len()
    }
}

pub trait StreamPosition {
    /// Gets the current position in the stream.
    fn stream_position(&self) -> usize;
}

/// A passthrough `Write` implementation that keeps track of position.
///
/// This is used to keep track of the current position in the output stream, since we can't use
/// `Seek` on all output streams. Additionally this allows us to keep track of the position
/// when using functions like `copy` that may call write several times and potentially fail.
pub struct StreamPositionTracker<T: Write> {
    writer: T,
    pos: usize,
}

impl<T: Write> StreamPositionTracker<T> {
    pub fn new(writer: T) -> Self {
        Self { writer, pos: 0 }
    }
}

impl<T: Write> Write for StreamPositionTracker<T> {
    /// Passthrough to the underlying writer, but also updates the position.
    fn write(&mut self, buf: &[u8]) -> io::Result<usize> {
        let written = self.writer.write(buf)?;
        self.pos += written;
        Ok(written)
    }

    // Passthrough to the underlying writer.
    fn flush(&mut self) -> io::Result<()> {
        self.writer.flush()
    }

    fn write_vectored(&mut self, bufs: &[IoSlice<'_>]) -> std::io::Result<usize> {
        let written = self.writer.write_vectored(bufs)?;
        self.pos += written;
        Ok(written)
    }
}

impl<T: Write> StreamPosition for StreamPositionTracker<T> {
    fn stream_position(&self) -> usize {
        self.pos
    }
}

/// Stream wrapper that implements a subset of `Seek`.
///
/// This is needed because we read from streams that do not implement `Seek`,
/// such as `Stdin`. Instead of implementing ad-hoc skipping we'll implement
/// `Seek` such that it only allows seeking forward. Any seek from the end of
/// the stream, or that would go backwards, will result in an error.
pub struct ForwardOnlySeeker<T: Read> {
    reader: T,
    pos: u64,
}

impl<T: Read> ForwardOnlySeeker<T> {
    pub fn new(reader: T) -> Self {
        Self { reader, pos: 0 }
    }
}

impl<T: Read> Read for ForwardOnlySeeker<T> {
    /// Passthrough to the underlying reader, but also updates the position.
    fn read(&mut self, buf: &mut [u8]) -> io::Result<usize> {
        let bytes_read = self.reader.read(buf)?;
        self.pos += bytes_read as u64;
        Ok(bytes_read)
    }
}

impl<T: Read> Seek for ForwardOnlySeeker<T> {
    /// Seeks forward in the stream, returns an error if seeking backwards.
    fn seek(&mut self, pos: SeekFrom) -> io::Result<u64> {
        let seek_size = match pos {
            SeekFrom::Current(offset) => u64::try_from(offset).ok(),
            SeekFrom::Start(offset) => offset.checked_sub(self.pos),
            SeekFrom::End(_) => {
                return Err(io::Error::new(
                    ErrorKind::InvalidInput,
                    "Cannot seek from end",
                ))
            }
        };

        match seek_size {
            Some(seek_size) => {
                copy(&mut self.by_ref().take(seek_size), &mut sink())?;
                Ok(self.pos)
            }
            None => Err(io::Error::new(
                ErrorKind::InvalidInput,
                "Only seeking forward allowed",
            )),
        }
    }

    fn stream_position(&mut self) -> io::Result<u64> {
        Ok(self.pos)
    }
}

#[cfg(test)]
mod test {
    use rstest::rstest;

    use super::*;

    #[test]
    fn test_write_cursor() {
        let mut buf = Vec::new();
        let mut writer = StreamPositionTracker::new(&mut buf);

        writer.write_all(b"hello").unwrap();
        assert_eq!(writer.stream_position(), 5);
        writer.write_all(b" world").unwrap();
        assert_eq!(writer.stream_position(), 11);
        writer.write_all(b"!").unwrap();
        assert_eq!(writer.stream_position(), 12);

        assert_eq!(buf, b"hello world!");
    }

    #[test]
    fn test_write_vectored_cursor() {
        let mut buf = Vec::new();
        let mut writer = StreamPositionTracker::new(&mut buf);

        let write_vector = [
            IoSlice::new(b"hello"),
            IoSlice::new(b" world"),
            IoSlice::new(b"!"),
        ];
        let bytes_written = writer.write_vectored(&write_vector).unwrap();

        assert_eq!(bytes_written, 12);
        assert_eq!(writer.stream_position(), 12);
        assert_eq!(buf, b"hello world!");
    }

    #[test]
    fn test_forward_seeker_stream() {
        let mut input_stream = b"Hello world".as_ref();

        let mut reader = ForwardOnlySeeker::new(&mut input_stream);
        let mut out_buf = [0u8; 5];
        reader.read_exact(&mut out_buf).unwrap();
        assert_eq!(&out_buf, b"Hello");

        reader.seek(SeekFrom::Current(1)).unwrap();
        reader.read_exact(&mut out_buf).unwrap();
        assert_eq!(&out_buf, b"world");
    }

    #[rstest]
    #[case(SeekFrom::End(0))]
    #[case(SeekFrom::Current(-1))]
    #[case(SeekFrom::Start(0))]
    fn test_forward_seeker_seek_fail(#[case] seek: SeekFrom) {
        let mut reader = ForwardOnlySeeker::new(b"Hello world".as_ref());
        reader.seek(SeekFrom::Start(1)).unwrap();
        assert!(reader.seek(seek).is_err());
    }
}
