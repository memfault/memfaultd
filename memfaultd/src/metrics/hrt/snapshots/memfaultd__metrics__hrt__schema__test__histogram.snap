---
source: memfaultd/src/metrics/hrt/schema.rs
expression: json!(datum)
---
{
  "t": 1234,
  "value": 123.0
}
